identity-iota/src/tangle/publish.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use identity_did::verification::MethodRef;
+use identity_did::verification::MethodType;
+use identity_did::verification::VerificationMethod;
+
+use crate::did::IotaDocument;
+
+// Method types allowed to sign a DID document update.
+pub const UPDATE_METHOD_TYPES: &[MethodType] = &[MethodType::Ed25519VerificationKey2018];
+
+/// Determines whether an updated document needs to be published as an integration or diff message.
+#[derive(Clone, Copy, Debug)]
+pub enum PublishType {
+  Integration,
+  Diff,
+}
+
+impl PublishType {
+  /// Compares two versions of a document and returns whether it needs to be published
+  /// as an integration or diff message. If `None` is returned, no update is required.
+  ///
+  /// Note: A newly created document must always be published as an integration message, and
+  /// this method does not handle this case.
+  pub fn new(old_doc: &IotaDocument, new_doc: &IotaDocument) -> Option<PublishType> {
+    if old_doc == new_doc {
+      return None;
+    }
+
+    let old_capability_invocation_set: Vec<Option<&VerificationMethod>> = old_doc
+      .as_document()
+      .capability_invocation()
+      .iter()
+      .map(|method_ref| match method_ref {
+        MethodRef::Embed(method) => Some(method),
+        MethodRef::Refer(did_url) => old_doc.as_document().resolve_method(did_url),
+      })
+      .filter(|method| {
+        if let Some(method) = method {
+          UPDATE_METHOD_TYPES.contains(&method.key_type())
+        } else {
+          true
+        }
+      })
+      .collect();
```

Comment:
Nit: this _could_ be made its own function e.g. `extract_signing_keys` since its a bit long and it's important that the exact same procedure is used for both documents. Completely optional though.